connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java

```diff
@@ -210,82 +208,88 @@ public void putTaskConfigs(String connName, List<Map<String, String>> configs, C
+        for (ConnectorTaskId taskId : configState.tasks(connName)) {
+            TaskConfig config = new TaskConfig(configState.taskConfig(taskId));
             try {
-                worker.startTask(taskId, config, this);
+                worker.startTask(taskId, config, this, initialState);
             } catch (Throwable e) {
                 log.error("Failed to add task {}: ", taskId, e);
                 // Swallow this so we can continue updating the rest of the tasks
                 // FIXME what's the proper response? Kill all the tasks? Consider this the same as a task
                 // that died after starting successfully.
             }
-            index++;
         }
     }
 
-    private Set<ConnectorTaskId> tasksFor(ConnectorState state) {
-        Set<ConnectorTaskId> tasks = new HashSet<>();
-        for (int i = 0; i < state.taskConfigs.size(); i++)
-            tasks.add(new ConnectorTaskId(state.name, i));
-        return tasks;
-    }
-
     private void removeConnectorTasks(String connName) {
-        ConnectorState state = connectors.get(connName);
-        Set<ConnectorTaskId> tasks = tasksFor(state);
+        Collection<ConnectorTaskId> tasks = configState.tasks(connName);
         if (!tasks.isEmpty()) {
             worker.stopTasks(tasks);
             worker.awaitStopTasks(tasks);
-            state.taskConfigs = new ArrayList<>();
+            configBackingStore.removeTaskConfigs(connName);
         }
     }
 
     private void updateConnectorTasks(String connName) {
-        List<Map<String, String>> newTaskConfigs = recomputeTaskConfigs(connName);
-        ConnectorState state = connectors.get(connName);
-        if (!newTaskConfigs.equals(state.taskConfigs)) {
+        if (configState.targetState(connName) == TargetState.PAUSED) {
+            log.trace("Skipping reconfiguration of tasks for paused connector {}", connName);
+            return;
+        }
+
+        Map<ConnectorTaskId, Map<String, String>> newTaskConfigs = recomputeTaskConfigs(connName);
+        Map<ConnectorTaskId, Map<String, String>> oldTaskConfigs = configState.allTaskConfigs(connName);
+
+        if (!newTaskConfigs.equals(oldTaskConfigs)) {
             removeConnectorTasks(connName);
-            state.taskConfigs = newTaskConfigs;
-            createConnectorTasks(connName);
+            configBackingStore.putTaskConfigs(connName, newTaskConfigs);
+            createConnectorTasks(connName, configState.targetState(connName));
         }
     }
 
+    private class ConfigUpdateListener implements ConfigBackingStore.UpdateListener {
 
-    private static class ConnectorState {
-        public String name;
-        public Map<String, String> configOriginals;
-        public ConnectorConfig config;
-        List<Map<String, String>> taskConfigs;
+        @Override
+        public void onConnectorConfigUpdate(String connector) {
+            configState = configBackingStore.snapshot();
```

Comment:
Hmm, is there anything to be done here? I think previously we just weren't considering changes to connector/task configs because standalone mode hadn't really been designed with that in mind, although I guess technically there's no reason you couldn't manage connectors via the REST API instead of (or in addition to) passing in config files. Right now it looks like we're in a sort of intermediate state where we could get these updates but we wouldn't actually apply them? Maybe to avoid complicating this patch we can just note this and maybe make it a TODO?